keps/sig-api-machinery/20190723-coordinated-discovery-document.md

```diff
@@ -0,0 +1,181 @@
+
+When responding to discovery document requests, apart from the original
+discovery documents, the API server also sets a `Consistent` field to indicate if
+all API servers are serving the same discovery documents, i.e., if all hash
+values recorded in `DiscoveryDocHashes` are equal. 
+
+## API changes
+
+We introduce a new API `DiscoveryDocHashes`, in a new API group
+`APIServerCoordination.k8s.io`. 
+
+```golang
+type DiscoveryDocHashes struct {
+   Hashes []Hash
+}
+
+type Hash struct {
+   // The ID of the reporting API server. 
+   // The ID can be generated in a similar fashion as the controller manager
+   // generates leader election id at
+   // https://github.com/kubernetes/kubernetes/blob/2321d1e9e8950cc94b3ef2368dfaacec61f1ba4f/cmd/cloud-controller-manager/app/controllermanager.go#L185.
+   // Required.
+   APIServerID string
+
+   // A hash calculated based on the discovery documents of built-in APIs. If
+   // two hashes equal, it means that the two API servers serve the same set of
+   // APIs.
+   // Required.
+   BuiltInAPIHash string
+
+   // LastHeartbeatTime is the last time the reporting API server updates this 
+   // field.
+   // Required 
+   LastHeartbeatTime metav1.Time
+}
+```
+
+API server instances report the hashes of their discovery documents through this
+API.  Although the main users of this API are the API servers, we make this API
+public to make debugging convenient.
+
+Another API change is that at each level of the discovery [API][] (APIGroupList,
+APIVersions, and APIResourceList), we add a new field `Consistent` to indicate
+if all API servers agree on the discovery documents, i.e., if all hashes
+reported in `DiscoveryDocHashes` are equal.
+
+Alternative to adding the `Consistent` field, we have considered letting API
+servers respond 500 error to discovery requests if there are multiple hash
+values in `DiscoveryDocHashes`. This is less desirable as it breaks old clients.
+
+## Changes to API servers
+
+### Calculating DiscoveryDocHashes
+
+Every API server instance calculates the DiscoveryDocHashes based on its in
+memory discovery documents for **built-in** resources. One API server instance
+only generates one DiscoveryDocHashes. And because built-in resources are
+static, the hash does not change through the life time of the API server
+process.
+
+The hash does not include the CRDs, because API server instances *can* serve
+consistently if using the `CRD` API as a coordination point. For example, when
+serving a CR request, if the API server always uses a [handler][] built based on
+the latest `CRD` stored in etcd, then all API server instances will behave
+consistently.
+
+[handler]:https://github.com/kubernetes/kubernetes/blob/63a43402a365f7a7615e01ea3e174fb8a71d67a8/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L200
+
+Similarly, for aggregated APIs, the API servers can use the `APIService` API as
+a coordination point to achieve consistency.
```

Comment:
This doesn't work because the service backing the APIService can change without any corresponding change in the APIService object.